enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,463 @@
+  - As a bonus, OLM will not increase the scope of APIs that it owns.
+
+### How Operators will opt-in
+
+Operators must opt-in to "OLM Supported Conditions" by adding an annotation to their owned CRD, where the key is the defined by OLM and the value is the `type` of the condition on the CR.
+
+#### Opt-in Example
+
+The "OLM Supported Condition" annotation key will adhere to the following format: `olm.Condition.<ConditionName>`. The value of the annotation will match the `Type` of the condition.  For example, the `NotInterruptible` condition's annotation key will be `olm.Condition.NotInterruptible`.
+
+While most "OLM Supported Condition" annotation keys will map to a single condition `type`, the `Important` condition will map to multiple condition types separated by commas.
+
+Let's consider an operator that installs a database. Assume that the operator already communicates that a database migration is in progress by setting the "Migrating" condition type in the status of the `foo` CustomResource. In this case, the operator should update the CRD to include the following annotation:
+
+```yaml
+"metadata": {
+  "annotations": {
+    "olm.Condition.NotInterruptible" : "Migrating",
+    "olm.Condition.Important" : "badConnectivity,unhealthyDatabase",
+  }
+}
+```
+
+By adding this annotation, OLM will search for all `foo` CustomResources and search for the `Migrating` Condition before upgrading the operator to the latest version.
+
+Now let's view the `foo` CR on cluster:
+
+```yaml
+apiVersion: foo.bar.com/v1
+kind: Foo
+metadata:
+  name: foo-database
+spec:
+  ...
+  ...
+  ...
+status:
+  conditions:
+  - type: Migrating # Condition 1
+    status: true
+    reason: "CriticalProcess" 
+    message: "Migrating old database schema to new format"
+  - type: unhealthyDatabase # Condition 2
+    status: true
+    reason: "LivenessProbeFailure"
+    message: "The foo-database failed its liveness probe."
+  - type: badConnectivity # Condition 2
+    status: true
+    reason: "404" 
+    message: "Unable to reach the database"
+```
+
+From the presence of the `Migrating` condition, OLM will understand that it should not upgrade the operator. The Operator also communicates that the `reason` the operator cannot be upgraded is due to a "critical process" and provides a human readable `message` for cluster admins.
+
+OLM will also highlight Conditions 1 and 2 within the new [Operator API](https://github.com/operator-framework/api/blob/master/crds/operators.coreos.com_operators.yaml) within the [condtions](https://github.com/operator-framework/api/blob/master/crds/operators.coreos.com_operators.yaml#L108) in the status.
+
+Note that by adding the `NotInterruptible` "OLM Supported Condition" to each CR, OLM will not upgrade the operator if any CR reports this condition as `true`.
```

Comment:
This can block operator upgrades if a particular CR ends up in a bad state and is unable to recover (for example the DB migration fails). I think it's valid to block the upgrade in that case regardless but its worth pointing out - at that point only a manual cleanup will allow for an upgrade.

Reply:
Suggestion here to help UX in this scenario: https://github.com/operator-framework/enhancements/pull/23#discussion_r441609005

Reply:
I think requiring cluster admin intervention is the safest approach, but @jupierce offered an interesting [alternative](https://github.com/operator-framework/enhancements/pull/23/files#r441609005).

Reply:
I think that's a very interesting idea, to have hard and soft limits for the condition and proceeding anyway. Some thoughts around this:

- By mandating a certain status gets updated every 30 minutes we are really starting to mandate how operator authors write their operator, which lowers adoption. Asking them to write a certain condition to their status is straightforward enough (they are doing it anyway) but saying to do it periodically is different. The logic definitely grows more complex, and I think we would be the ones to have to provide a library for them to use. 
- Conceptually, a grace period is used when something is wrong. I don't think `NotInterruptible` is necessarily a bad condition that an operator needs to get out of - it's a transient condition that indicates a workload is in progress and has state, like an ML training job. We would expect that eventually the workload comes out of that state and the operator goes back to an upgradable state. 
- as @awgreene mentioned, OLM terminating workloads without user input is pretty risky behavior in general. Any deletion/upgrade event is initiated by the user, either by deleting the CSV or the subscription moving to a new version. The subscription makes the process seamless but still the user updates the catalog on the cluster with the new version, either explicitly or via catalog polling. It's really up to the operator author to make sure their operands are well behaved. 

All that being said, I think you suggestion brings up a need for an example, in code, of how to set and unset the condition and best practices for it. That way operator authors have a clear picture of how to use the condition to their benefit.